fix TAF tangent-linear bits for mixing free and fixed format source code

mjlosch opened the pull request:
## What changes does this PR introduce?
feature completion

## What is the current behaviour? 
tangent linear TAFization is restricted to fixed format source files, see #855 and PR #709

## What is the new behaviour 
- in analogy to PR #709 fix taf-related tangent linear targets in `genmake2`
- modified `verification/hs94.1x64x5` to use free format code for AD and TLM
- replace `*.list` by `*_ad_diff.list.`

## Does this PR introduce a breaking change? 
pure f77-compilers (e.g. g77) may not be able to compile AD version of `verification/hs94.1x64x5`

## Other information:
o additionally, one could enforce TAF_FORTRAN_VERS > F77 and include more F90 features in new file `cost_test_local.F90`

## Suggested addition to `tag-index`
o modify verification/hs94.1x64x5 to use free format code for AD and TLM
o tools/genmake2:
  - change Makefile to enable TAF to generate tangent linear code from a mix of
     fixed-form (.F) and free-form (.F90) src code.
  - replace *.list by *_ad_diff.list to avoid accidents

jm-c reviewed (commented):
I made few tille adjustments:

1. The definition of "_RS" in `cost_test_local.F90` is fixed (I think). However, I don't think it's used in this file, so it could be removed ?
2. I did not check, but I think for the Adjoint test that I am running on svante with pgi compiler would need the `genmake_local` to compile, so I added it here.
3. the 2 blank lines I added in `genmake2` are not necessary, but easier to read the Makefile (+ similar to the AD pieces).

I am currently running few testreport ("-tlm", "-adm" with and without "-ncad") just to double check, but overall this PR looks good.

One detail: I would suggest to remove the comments 
>  unused target, could be removed

just before "adobj" and "ftlobj" tagets: There is a similar "obj" target for forward-only set of objet-files that might be used as intermediate step for "lib" (libmitgcmuv.a).
This library option/target is not in place for AD or TL code, but since these 2 targets  "adobj" and "ftlobj" are relatively simple and easy to maintain, we could keep them without this comment in the Makefile.

jm-c commented:
Not directly related to this PR: It's not clear to me why we treat the "\*.flow" files differently for AD and TL:
For AD, they all go through the CPP pre-compiler (change from "\*.flow" to "\*.flowdir") and we use the updated list of flow-directive files "FLOWFILES".
But for TL, we use instead the original set of flow directive files "AD_FLOW_FILES" ("\*.flow") without going through CPP step. Is there a good reason for that ?

Now assuming there is a good reason to keep it this way, I don't think we need this line:
>  @\$(MAKE) -f \$(MAKEFILE) \$(AD_FLOW_FILES)

(line 3596 in master branch version of `genmake2`) for the "ftl_input_code.$FS" target, because this line does not do anything, I think.

mjlosch commented:
> Not directly related to this PR: It's not clear to me why we treat the "*.flow" files differently for AD and TL:

I have no idea, I will add this to the list #855

jm-c commented on verification/hs94.1x64x5/code_ad/cost_test_local.F90:
```diff
@@ -0,0 +1,87 @@
+!     !INTERFACE:
+subroutine cost_test_local ( &
+     sNx, sNy, nSx, nSy, OLx, OLy, Nr, &
+     myByLo, myByHi, myBxLo, myBxHi, &
```
```suggestion
     myBxLo, myBxHi, myByLo, myByHi, &
```
since the usual order is X first and then Y (like just above: "sNx, sNy").

mjlosch commented:
> 1. The definition of "_RS" in `cost_test_local.F90` is fixed (I think). However, I don't think it's used in this file, so it could be removed ?

Yes, can be removed, and I will do it (left over from previous verification experiment)
> 2. I did not check, but I think for the Adjoint test that I am running on svante with pgi compiler would need the `genmake_local` to compile, so I added it here.

Good, that means that we can use more F90 syntax in `cost_test_local.F90`. I will look into that.
> 3. the 2 blank lines I added in `genmake2` are not necessary, but easier to read the Makefile (+ similar to the AD pieces).

OK

> One detail: I would suggest to remove the comments
> 
> > unused target, could be removed
> 
> just before "adobj" and "ftlobj" tagets: There is a similar "obj" target for forward-only set of objet-files that might be used as intermediate step for "lib" (libmitgcmuv.a). This library option/target is not in place for AD or TL code, but since these 2 targets "adobj" and "ftlobj" are relatively simple and easy to maintain, we could keep them without this comment in the Makefile.

OK, done. Should we remove the targets `adtafonly` and `ftltafonly`?

jm-c reviewed (approved):
Looks good. Will check that the modified experiment run well with PGI on svante.

jm-c commented:
And forgot about this:
> Should we remove the targets adtafonly and ftltafonly?

I think these would fit better in one of the "cleaning" PR such as #864 or #866

jm-c commented:
I checked that the modified exp. run OK with PGI on svante. It turns out that the `genmake_local` was not necessary for compiling (since the "\*.F90" files are compiled with "F90C") but was needed to make the executable when not using MPI (probably with MPI all the F90 libs are already there in "mpif77").
Anyway, I think this PR is ready to be merged and will likely do that tomorrow unless someone wants to add something. @mjlosch do you agree ?

mjlosch commented:
Great, I don't have anything to add.
